### PR TITLE
fix(image): dataUri is not random

### LINKED ISSUE #2307: faker.image.dataUri is not random
faker.image.dataUri is not random, it always returns the same image (a grey rectangle with the provided width/height in text)

https://stackblitz.com/edit/faker-js-demo-quium5?file=index.ts

Source: https://github.com/faker-js/faker/blob/651d1a8c/src/modules/image/index.ts#L373

I also noted that the width x height text doesn't fit for small `width` parameters

---

_Originally posted by @ST-DDT in https://github.com/faker-js/faker/pull/2273#discussion_r1290432388_

### PR DESCRIPTION
**Pull request by OmkarPednekar**
If color option is not provided the code will use a randomly generated color code as default color
the fix(module):faker.image.dataUri commit is the final one. where I have used faker.color.rgb in the previous commit I was using a for loop to generate a hex code. @ST-DDT suggested me to use the inbuilt functions so that we have reproducible values.

fixes #2307

### CODE REVIEW

File: src/modules/image/index.ts [hunk 1]
```diff
@@ -377,12 +377,16 @@ export class ImageModule {
...
+       * @default 'random color'
        */
       color?: string;
     } = {}
   ): string {
-    const { width = 640, height = 480, color = 'grey' } = options;
+    const {
+      width = 640,
+      height = 480,
+      color = this.faker.color.rgb({ prefix: '#' }),
```
Comment (ST-DDT) [inline src/modules/image/index.ts]:
```suggestion
      color = this.faker.color.rgb(),
```

File: src/modules/image/index.ts [hunk 2]
```diff
@@ -377,12 +377,16 @@ export class ImageModule {
       /**
        * The color of the image.
        *
-       * @default 'grey'
+       * @default 'random color'
```
Comment (matthewmayer) [inline src/modules/image/index.ts]: this makes it look as if the default value is the literal string `'random color'`. I'd either change it to 

`
@default a random color
`

or omit the default value completely, and mention in the description that the color will be random unless specified.

Comment (OmkarPednekar) [reply src/modules/image/index.ts]: Ok, will do the changes

Comment (OmkarPednekar) [reply src/modules/image/index.ts]: image
 is this okay?

**Comment by ST-DDT** [comment]
I just noticed, that `dataUri` doesn't have seeded tests, so I created a PR for that.

Blocked by:

- https://github.com/faker-js/faker/pull/2317

**Comment by OmkarPednekar** [comment]
is this okay? 
image

**Comment by ST-DDT** [comment]
> is this okay? image

I would probably phrase it like this:

`The background color of the image. Must be a color supported by svg. Defaults to a random color.`

**Comment by OmkarPednekar** [comment]
Okay

File: src/modules/image/index.ts (hunk 2 above)
Comment (ST-DDT) [reply src/modules/image/index.ts]: Other methods use:

````ts
@default faker.color.rgb()
````

**Comment by OmkarPednekar** [comment]
Please Check

File: src/modules/image/index.ts [hunk 3]
```diff
@@ -375,17 +375,16 @@ export class ImageModule {
        */
       height?: number;
       /**
-       * The color of the image.
+       * The color of the image. Must be a color supported by svg. Defaults to a random color.
```
Comment (ST-DDT) [inline src/modules/image/index.ts]: Don't forget to change the `@param` docs above.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

**Comment by OmkarPednekar** [comment]
Made the changes in @param as well

File: src/modules/image/index.ts (hunk 3 above)
Comment (OmkarPednekar) [reply src/modules/image/index.ts]: Done

**Comment by OmkarPednekar** [comment]
image
Updated the Docs

File: src/modules/image/index.ts [hunk 4]
```diff
@@ -353,7 +353,7 @@ export class ImageModule {
    * @param options Options for generating a data uri.
    * @param options.width The width of the image. Defaults to `640`.
    * @param options.height The height of the image. Defaults to `480`.
-   * @param options.color The color of the image. Defaults to `grey`.
+   * @param options.color The color of the image. Defaults to a random color.
```
Comment (ST-DDT) [inline src/modules/image/index.ts]: Both the `@param` description and the field description should be the same/similar.

```suggestion
   * @param options.color The color of the image. Must be a color supported by svg. Defaults to a random color.
```

File: src/modules/image/index.ts [hunk 5]
```diff
@@ -375,14 +375,18 @@ export class ImageModule {
        */
       height?: number;
       /**
-       * The color of the image.
+       * The color of the image. Must be a color supported by svg. Defaults to a random color.
```
Comment (ST-DDT) [inline src/modules/image/index.ts]:
```suggestion
       * The color of the image. Must be a color supported by svg.
```

The `@default` already covers the `Defaults to` part.

File: test/modules/image.spec.ts [hunk 6]
```diff
@@ -471,8 +471,9 @@ describe('image', () => {
 
   describe('dataUri', () => {
     it('should return a blank data', () => {
+      const expectedPattern = /^data:image\/svg\+xml;charset=UTF-8,/;
       const dataUri = faker.image.dataUri({ width: 200, height: 300 });
-      expect(dataUri).toMatchSnapshot();
+      expect(dataUri).toMatch(expectedPattern);
```
Comment (ST-DDT) [inline test/modules/image.spec.ts]: Please inline the expected pattern.

Comment (ST-DDT) [reply test/modules/image.spec.ts]: Maybe also check that the string contains the given width and height.

File: test/modules/image.spec.ts [hunk 7]
```diff
@@ -471,8 +471,9 @@ describe('image', () => {
...
     });
 
     it('should return a background color data URI', () => {
```
Comment (ST-DDT) [inline test/modules/image.spec.ts]: Maybe use the `expectedPattern` like check here as well (in addition to the snapshot)?

File: src/modules/image/index.ts (hunk 5 above)
Comment (OmkarPednekar) [reply src/modules/image/index.ts]: okay

File: test/modules/image.spec.ts (hunk 7 above)
Comment (OmkarPednekar) [reply test/modules/image.spec.ts:479]: in this testcase 'should return a background color data URI' ?

Comment (OmkarPednekar) [reply test/modules/image.spec.ts]: image
Made changes to the description and also added the testcase for checking width and height to both tests

File: test/modules/image.spec.ts [hunk 8]
```diff
@@ -481,6 +482,8 @@ describe('image', () => {
         height: 300,
         color: 'red',
       });
+      expect(dataUri).toMatch(/^data:image\/svg\+xml;charset=UTF-8,/);
+      expect(dataUri).toMatch(/width%3D%22200%22%20height%3D%22300/);
```
Comment (ST-DDT) [inline test/modules/image.spec.ts:519]: Don't forget to check for the color.

Comment (OmkarPednekar) [reply test/modules/image.spec.ts:519]: Added the check for color.

File: src/modules/image/index.ts (hunk 4 above)
Comment (ST-DDT) [reply src/modules/image/index.ts]: Looks like you forgot this one. Or did you mark this as resolved intentionally?

Comment (OmkarPednekar) [reply src/modules/image/index.ts]: Oh! So sorry I totally missed this one

Comment (OmkarPednekar) [reply src/modules/image/index.ts]: The color of the image. Must be a color supported by svg. it should be this right? same as description

Comment (ST-DDT) [reply src/modules/image/index.ts]: Plus the `Defaults to ...` part because that's not present in the `@param` block.

**Review by ST-DDT — DISMISSED** [review]

**Comment by OmkarPednekar** [comment]
any updates on the review @matthewmayer

**Comment by matthewmayer** [comment]
It's quite normal that PRs will take weeks or months before merging even if they are fine. Don't worry 😀

**Comment by OmkarPednekar** [comment]
Okay Got It and also why is this one test failing?

**Comment by ST-DDT** [comment]
> Okay Got It and also why is this one test failing?

Fixed

**Comment by ST-DDT** [comment]
We added some (forgotten) seeded tests to ensure the values generated by this method are reproducible.
Please run `pnpm run preflight` to update the values.

**Comment by OmkarPednekar** [comment]
> We added some (forgotten) seeded tests to ensure the values generated by this method are reproducible. Please run `pnpm run preflight` to update the values.

So i have to do this and then  commit again?

**Comment by OmkarPednekar** [comment]
why are these checks failing?

**Comment by ST-DDT** [comment]
> why are these checks failing?

The (new) seeded tests should highlight changed results caused by a refactoring/change.
This PR changes the behavior of the method (slightly) thus the  test snapshots needed to be updated.

**Review by ST-DDT — DISMISSED** [review]

**Review by xDivisionByZerox — DISMISSED** [review]

**Comment by xDivisionByZerox** [comment]
@OmkarPednekar Please resolve the merge conflicts if possible.

**Review by ejcheng — DISMISSED** [review]

**Review by ST-DDT — DISMISSED** [review]

**Comment by ST-DDT** [comment]
I fixed the merge conflicts myself to speed up the merge process.

**Review by Shinigami92 — DISMISSED** [review]

**Review by ST-DDT — DISMISSED** [review]

**Review by Shinigami92 — DISMISSED** [review]

**Comment by Shinigami92** [comment]
👀

**Comment by OmkarPednekar** [comment]
Why are these tests failing?

**Comment by ST-DDT** [comment]
I fixed the random seeded tests using test snapshots.

**Comment by ST-DDT** [comment]
> Why are these tests failing?

Because the randomized tests used snapshots which would work fine if the result is static, but this PR changes them to be random thus failing the tests. I fixed them by not using the static snapshots and instead check via validatorjs's isDataUri.

**Review by ST-DDT — APPROVED** [review]

**Comment by OmkarPednekar** [comment]
> > Why are these tests failing?
> 
> Because the randomized tests used snapshots which would work fine if the result is static, but this PR changes them to be random thus failing the tests. I fixed them by not using the static snapshots and instead check via validatorjs's isDataUri.

But I had removed those test cases and changed them to match the changes

**Comment by ST-DDT** [comment]
> But I had removed those test cases and changed them to match the changes

https://github.com/faker-js/faker/pull/2273 added new ones.

- https://github.com/faker-js/faker/pull/2273

**Comment by OmkarPednekar** [comment]
> > But I had removed those test cases and changed them to match the changes
> 
> #2273 added new ones.
> 
> * [feat(image): add image dataUri with base64 #2273](https://github.com/faker-js/faker/pull/2273)

Oh I seee so how can we fix this failing tests?

**Comment by ST-DDT** [comment]
> Oh I seee so how can we fix this failing tests?

I already removed/fixed them.

**Comment by OmkarPednekar** [comment]
> > Oh I seee so how can we fix this failing tests?
> 
> I already removed/fixed them.

No I meant the merge checks

**Comment by ST-DDT** [comment]
> No I meant the merge checks

Not sure what you are referring to.

grafik

**Review by Shinigami92 — APPROVED** [review]

**Comment by OmkarPednekar** [comment]
> > No I meant the merge checks
> 
> Not sure what you are referring to.
> 
> grafik

Oh whatt just a moment ago they were appearing failed lol

**Pull request merged**